### PR TITLE
Fix treasury minting

### PR DESCRIPTION
**Pull request by gavofyork**
Minting is meant to be such that correctly functioning stakers don't suffer dilution. This fixes that by reporting to the treasury (which mints funds based upon what stakers have been given in return for what stake) not just the amount minted, but also the portion of the stake that it was minted in as a reward for.

Given that not all validators have the same amount at stake, we use the *most staked validator* multiplied by the number of validators as the latter argument. This ensures that the least relatively rewarded validator-group will not suffer overall dilution.

### CODE REVIEW

File: substrate/runtime/treasury/src/lib.rs [hunk 1]
```diff
@@ -278,9 +278,13 @@ impl<T: Trait> Module<T> {
 	}
 }
 
-impl<T: Trait> OnMinted<T::Balance> for Module<T> {
-	fn on_minted(b: T::Balance) {
-		<Pot<T>>::put(Self::pot() + b);
+impl<T: Trait> OnDilution<T::Balance> for Module<T> {
+	fn on_dilution(minted: T::Balance, portion: T::Balance) {
+		if !minted.is_zero() && !portion.is_zero() {
```
Comment (rphmeier) [inline substrate/runtime/treasury/src/lib.rs]: a comment here describing that extra tokens are minted to the treasury whenever new balances are minted would be useful

File: substrate/runtime/balances/src/lib.rs [hunk 2]
```diff
@@ -93,14 +93,14 @@ impl<
...
 }
 
-/// Trait for a hook to get called when some balance has been minted.
-pub trait OnMinted<Balance> {
-	/// Some balance `b` was minted.
-	fn on_minted(b: Balance);
+/// Trait for a hook to get called when some balance has been minted, causing dilution.
+pub trait OnDilution<Balance> {
+	/// Some `portion` of the total balance "grew" by `minted`.
+	fn on_dilution(minted: Balance, portion: Balance);
```
Comment (rphmeier) [inline substrate/runtime/balances/src/lib.rs]: is the `portion` before or after having balance `minted` to it?

Comment (rphmeier) [reply substrate/runtime/balances/src/lib.rs]: (documentation should answer this question)

**Comment by rphmeier** [comment]
Using the maximum would encourage nominators to pile onto the highest-staked validators but not for the lowest-staked validators to increase their stake.

**Comment by gavofyork** [comment]
your reward would be smaller if you did that. the more extreme the difference between highest and lowest (that you're willfully not backing despite trusting), then the more reward you're giving up.

**Comment by rphmeier** [comment]
Filed this concern as (https://github.com/paritytech/substrate/issues/708) to be addressed

**Pull request merged**